sdk/web-pubsub/web-pubsub-client/src/webPubSubClient.ts

```diff
@@ -0,0 +1,1121 @@
+    }
+
+    this.buildMessageRetryOptions(clientOptions);
+
+    return clientOptions;
+  }
+
+  private buildMessageRetryOptions(clientOptions: WebPubSubClientOptions): void {
+    if (!clientOptions.messageRetryOptions) {
+      clientOptions.messageRetryOptions = {};
+    }
+
+    if (
+      clientOptions.messageRetryOptions.maxRetries == null ||
+      clientOptions.messageRetryOptions.maxRetries < 0
+    ) {
+      clientOptions.messageRetryOptions.maxRetries = 3;
+    }
+
+    if (
+      clientOptions.messageRetryOptions.retryDelayInMs == null ||
+      clientOptions.messageRetryOptions.retryDelayInMs < 0
+    ) {
+      clientOptions.messageRetryOptions.retryDelayInMs = 1000;
+    }
+
+    if (
+      clientOptions.messageRetryOptions.maxRetryDelayInMs == null ||
+      clientOptions.messageRetryOptions.maxRetryDelayInMs < 0
+    ) {
+      clientOptions.messageRetryOptions.maxRetryDelayInMs = 30000;
+    }
+
+    if (clientOptions.messageRetryOptions.mode == null) {
+      clientOptions.messageRetryOptions.mode = "Fixed";
+    }
+  }
+
+  private buildRecoveryUri(): string | null {
+    if (this._connectionId && this._reconnectionToken && this._uri) {
+      const url = new URL(this._uri);
+      url.searchParams.append("awps_connection_id", this._connectionId);
+      url.searchParams.append("awps_reconnection_token", this._reconnectionToken);
+      return url.toString();
+    }
+    return null;
+  }
+
+  private getOrAddGroup(name: string): WebPubSubGroup {
+    if (!this._groupMap.has(name)) {
+      this._groupMap.set(name, new WebPubSubGroup(name));
+    }
+    return this._groupMap.get(name) as WebPubSubGroup;
+  }
+
+  private ChangeState(newState: WebPubSubClientState): void {
+    console.debug(
+      `The client state transfer from ${this._state.toString()} to ${newState.toString()}`
+    );
+    this._state = newState;
+  }
+
+  private async operationExecuteWithRetry<T>(
+    inner: () => Promise<T>,
+    signal?: AbortSignalLike
+  ): Promise<T> {
+    let retryAttempt = 0;
+
+    while (true) {
+      try {
+        await inner.call(this);
+      } catch (err) {
+        retryAttempt++;
+        const delayInMs = this._messageRetryPolicy.nextRetryDelayInMs(retryAttempt);
+        if (delayInMs == null) {
+          throw err;
+        }
+
+        try {
+          await delay(delayInMs);
+        } catch {}
+
+        if (signal?.aborted) {
+          throw err;
+        }
+      }
+    }
+  }
+}
+
+class RetryPolicy {
+  private _retryOptions: WebPubSubRetryOptions;
+  private _maxRetriesToGetMaxDelay: number;
+
+  public constructor(retryOptions: WebPubSubRetryOptions) {
+    this._retryOptions = retryOptions;
+    this._maxRetriesToGetMaxDelay = Math.ceil(
+      Math.log2(this._retryOptions.maxRetryDelayInMs!) -
+        Math.log2(this._retryOptions.retryDelayInMs!) +
+        1
+    );
+  }
+
+  public nextRetryDelayInMs(retryAttempt: number): number | null {
```

Comment:
might be nice to factor this into a common `core-util` helper, since we do something similar here: https://github.com/Azure/azure-sdk-for-js/blob/8e52a3e66d9b8ac1e57b6844fc7cf1326e57cb1b/sdk/core/core-rest-pipeline/src/retryStrategies/exponentialRetryStrategy.ts#L69